pkg/controller/flow/reconcile.go

```diff
@@ -0,0 +1,360 @@
+
+	if err != nil {
+		return nil, err
+	}
+	newFlow.Status = flow.Status
+
+	// Until #38113 is merged, we must use Update instead of UpdateStatus to
+	// update the Status block of the Flow resource. UpdateStatus will not
+	// allow changes to the Spec of the resource, which is ideal for ensuring
+	// nothing other than resource status has been updated.
+	if err = r.client.Update(context.TODO(), newFlow); err != nil {
+		return nil, err
+	}
+	return newFlow, nil
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Flow resource
+// with the current status of the resource.
+func (r *reconciler) resolveActionTarget(namespace string, action v1alpha1.FlowAction) (string, error) {
+	glog.Infof("Resolving target: %+v", action)
+
+	if action.Target != nil {
+		return r.resolveObjectReference(namespace, action.Target)
+	}
+	if action.TargetURI != nil {
+		return *action.TargetURI, nil
+	}
+
+	return "", fmt.Errorf("No resolvable action target: %+v", action)
+}
+
+// resolveObjectReference fetches an object based on ObjectRefence. It assumes the
+// object has a status["domainInternal"] string in it and returns it.
+func (r *reconciler) resolveObjectReference(namespace string, ref *corev1.ObjectReference) (string, error) {
+	resourceClient, err := CreateResourceInterface(r.restConfig, ref, namespace)
+	if err != nil {
+		glog.Warningf("failed to create dynamic client resource: %v", err)
+		return "", err
+	}
+
+	obj, err := resourceClient.Get(ref.Name, metav1.GetOptions{})
+	if err != nil {
+		glog.Warningf("failed to get object: %v", err)
+		return "", err
+	}
+	status, ok := obj.Object["status"]
+	if !ok {
+		return "", fmt.Errorf("%q does not contain status", ref.Name)
+	}
+	statusMap := status.(map[string]interface{})
+	serviceName, ok := statusMap[targetFieldName]
+	if !ok {
+		return "", fmt.Errorf("%q does not contain field %q in status", targetFieldName, ref.Name)
+	}
+	serviceNameStr, ok := serviceName.(string)
+	if !ok {
+		return "", fmt.Errorf("%q status field %q is not a string", targetFieldName, ref.Name)
+	}
+	return serviceNameStr, nil
+}
+
+func (r *reconciler) reconcileChannel(flow *v1alpha1.Flow) (*channelsv1alpha1.Channel, error) {
+	channelName := flow.Name
+
+	channel := &channelsv1alpha1.Channel{}
+	err := r.client.Get(context.TODO(), client.ObjectKey{Namespace: flow.Namespace, Name: channelName}, channel)
+	if errors.IsNotFound(err) {
+		channel, err = r.createChannel(flow)
+		if err != nil {
+			glog.Errorf("Failed to create channel %q : %v", channelName, err)
+			return nil, err
+		}
+	} else if err != nil {
+		glog.Errorf("Failed to reconcile channel %q failed to get channels : %v", channelName, err)
+		return nil, err
+	}
+
+	// Should make sure channel is what it should be. For now, just assume it's fine
+	// if it exists.
+	return channel, err
```

Comment:
Seems like `err` can only be `nil` here, so it might be clearer to `return channel, nil`.

Reply:
done